Tools/Scripts/hooks/pre-push

```diff
@@ -0,0 +1,356 @@
+    if VERBOSITY > QUIET:
+        print('Attempting to push {} batch{} of commits to {}...'.format(len(to_push), '' if len(to_push) == 1 else 'es', name))
+
+    prompt_for = []
+    for commit, remotes in remotes_with.items():
+        security_level, categorization_class = security_level_for_commit(commit, remotes, name_to_remote=name_to_remote)
+
+        # Unkown security level of source
+        if categorization_class == 1 and security_level is None and not target_security_level:
+            print_error("'{}' comes from an uncategorized remote, it's not safe to push it publically".format(commit))
+            return 1
+        # Security level of source exceeds security level of target
+        if categorization_class == 1 and security_level > (target_security_level or 0):
+            print_error("'{}' comes from a more secure remote than '{}'".format(commit, name))
+            if MODE == PUBLISH_MODE:
+                prompt_for.append(commit)
+            else:
+                return 1
+
+        # Unknown security level from original of cherry-pick
+        if categorization_class == 2 and security_level is None and not target_security_level:
+            print_error("'{}' cherry-picks a change from an uncategorized remote, it's not safe to push it publically".format(commit))
+            return 1
+        # Security level of original of a cherry-pick exceeds security level of target
+        if categorization_class == 2 and security_level > (target_security_level or 0):
+            print_error("'{}' cherry-picks a change from a more secure remote than '{}'".format(commit, name))
+            if MODE == DEFAULT_MODE:
+                prompt_for.append(commit)
+            else:
+                return 1
+
+        # Commit is redacted, but we're pushing it publically
+        if categorization_class == 3 and (MAX_LEVEL if security_level is None else security_level) > (target_security_level or 0):
+            print_error("'{}' fixes a redacted issue, but '{}' is a public remote".format(commit, name))
+            if MODE == DEFAULT_MODE:
+                prompt_for.append(commit)
+            else:
+                return 1
+
+        if VERBOSITY >= VERBOSE:
+            print('    {} is security level {} by class {}'.format(commit, security_level, categorization_class))
+
+    if prompt_for and not TTY:
+        print('No TTY available for prompting user, rejecting push')
+        return 1
+
+    if prompt_for:
+        branches = sorted(list(set([branch_for(commit) for commit in prompt_for])))
+        expected_response = 'publicize {}'.format(
+            branches[0] if len(branches) == 1
+            else '{} and {}'.format(', '.join(branches[:-1]), branches[-1]),
+        )
+        print('')
+        print('We have detected you are about to publish security content to a public remote')
+        print('Unless you are engaged in merge-back or fixing a security bug which has not shipped, you are likely making a mistake')
```

Comment:
re "which has not shipped": How do contributors know this? If there's a silent "in Safari" in this sentence, then Apple engineers can determine whether a security bug has escaped, but I'm not aware of how to make that call more generally.

Reply:
I think the short answer here is that a contributor which does not understand how various WebKit ports ship code will be unable to answer this question. I think that's fine, such a contributor would then ask for help (which they would need to post a security PR anyways, since the repository where one does that is not publicly accessible).

The longer answer is that if you are fixing a security bug that has just been introduced in the last week or two, it's a good bet it hasn't shipped yet. Past that, answering the question "has this bug shipped" is a hard question to answer that requires coordination between various WebKit ports, and is not likely to be answerable with public information. While I have Safari in mind when writing this comment, making the statement general was deliberate, it applies to other shipping ports as well.